GPU build broken: cuKron_kernel fails for mixed ComplexFloat×Double (regression from #858)

## Summary

The GPU build of `master` is broken. Compiling `src/linalg/Kron.cpp` fails with a type error in the CUDA Kron kernel for mixed complex/real instantiations:

```
src/backend/linalg_internal_gpu/cuKron_internal.cuh(49): error: no operator "=" matches these operands
            operand types are: cuDoubleComplex = cuFloatComplex
          out[blockIdx.x * blockDim.x + threadIdx.x] = Lin[x] * Rin[y];

  detected during:
    instantiation of "void cytnx::linalg_internal::cuKron_kernel(TO *, ...) [with TO=cuDoubleComplex, TL=cuComplex, TR=cytnx::cytnx_double]"
    instantiation of "void cytnx::linalg_internal::cuKron_general(...) [with TO=cuDoubleComplex, TL=cuComplex, TR=cytnx::cytnx_double]" at line 90 of src/linalg/Kron.cpp
gmake[3]: *** [.../Kron.cpp.o] Error 1
```

## Root cause

Regression introduced by #858 (commit `004e433a`, "fix(Type): promote mixed complex/real dtypes by precision"). That change made mixed complex/real pairs promote by precision:

> ComplexFloat + Double now promotes to **ComplexDouble** (previously ComplexFloat).

`cuKron_general`/`cuKron_kernel` derive the output type via `type_promote_from_gpu_pointer_t<TL,TR>`, so for `(TL=ComplexFloat, TR=Double)` the output is now `cuDoubleComplex`. But the kernel computes `Lin[x] * Rin[y]` using the `operator*` overloads in `src/utils/cucomplex_arithmetic.cu`, where `cuFloatComplex * cytnx_double` returns **`cuFloatComplex`** (follows the left/complex operand's precision). Assigning that `cuFloatComplex` product into the `cuDoubleComplex` output has no matching `operator=`, hence the compile error.

| | `out_type` for (ComplexFloat, Double) | product `Lin*Rin` | result |
|---|---|---|---|
| before #858 | ComplexFloat | cuFloatComplex | compiles |
| after #858  | ComplexDouble | cuFloatComplex | **error** |

The broken instantiations are `(cuFloatComplex, cytnx_double)` and `(cytnx_double, cuFloatComplex)` (either operand order).

## Why CI didn't catch it

CI does not build the CUDA/GPU path (it needs a GPU + cuTENSOR), so the GPU-only template instantiation of Kron is never compiled. The CPU `Kron_general` path handles the conversion and compiles fine, so nothing flagged the regression when #858 merged.

Scope: only `src/linalg/Kron.cpp` uses `type_promote_from_gpu_pointer_t`, so this specific defect is isolated to the GPU Kron kernel. (Other GPU elementwise ops such as cuDiv already compute in the output type via a dispatch helper and are unaffected.)

## Reproduce

Configure with CUDA + cuTENSOR and build the `cytnx` target:

```
cmake -S . -B build -DUSE_CUDA=ON -DUSE_CUTENSOR=ON \
      -DCMAKE_CUDA_ARCHITECTURES=70   # any real arch
cmake --build build --target cytnx
```

The build fails at `Kron.cpp.o`.

## Suggested fix

Compute the product in the output type before assignment, mirroring the existing `cuDiv_dispatch.cu` pattern (`CuToComplexDouble`/`CuToComplexFloat` + `if constexpr` dispatch). i.e. replace

```cpp
out[...] = Lin[x] * Rin[y];
```

with a helper `cuKron_mul<TO>(Lin[x], Rin[y])` that promotes both operands to `TO` (`cuCmul`/`cuCmulf` for complex `TO`, `static_cast<TO>` otherwise). This makes the result assignable for every promoted `(TO, TL, TR)` triple.

## Environment

- CUDA 12.9, cuTENSOR 2.7.0, sm_70 (V100)
- master at the time of report; regression traces to #858 / `004e433a`

Posted by Claude Code on behalf of @pcchen

These are presumably broken as well, but I guess there are no tests?
```c++
Outer_ii[ComplexFloat][Double]
Outer_ii[Double][ComplexFloat]
cuAri_ii[ComplexFloat][Double]
cuAri_ii[Double][ComplexFloat]
cuOuter_ii[ComplexFloat][Double]
cuOuter_ii[Double][ComplexFloat]
```
eg
```c++
    // This function is generated by python script
    void cuOuter_internal_cftd(boost::intrusive_ptr<Storage_base> &out,
                               const boost::intrusive_ptr<Storage_base> &Lin,
                               const boost::intrusive_ptr<Storage_base> &Rin,
                               const cytnx_uint64 &j1, const cytnx_uint64 &j2) {
      cuOuter_general<cuFloatComplex, cuFloatComplex, cytnx_double>(out, Lin, Rin, j1, j2);
    }
```
needs to be
```diff
-cuOuter_general<cuFloatComplex, cuFloatComplex, cytnx_double>(out, Lin, Rin, j1, j2);
+cuOuter_general<cuDoubleComplex, cuFloatComplex, cytnx_double>(out, Lin, Rin, j1, j2);
```
But better solution would be to get rid of that dispatch mechanism entirely, it looks like those are the only cuda functions still using it.  The cuda kernel is not great itself.  `Outer` kernel seems a bit redundant anyway,
```c++
  Tensor Outer(const Tensor& a, const Tensor& b) {
    cytnx_error_msg(a.rank() != 1, "[ERROR] tensor #1 should have rank-1.%s", "\n");
    cytnx_error_msg(b.rank() != 1, "[ERROR] tensor #2 should have rank-1.%s", "\n");
    Tensor out = Kron(a, b);
    return out.reshape({a.shape()[0], b.shape()[0]});
  }
```

--- a/include/Tensor.hpp
+++ b/include/Tensor.hpp
@@ -514,9 +514,10 @@ namespace cytnx {
 
     // Convert storage to the logical Cytnx element pointer type.
     //
-    // Use ptr_as<T>() for host code and raw memory copies, including cudaMemcpy. For complex
-    // tensors, T is std::complex<...>. Use gpu_ptr_as<T>() only for CUDA APIs that require CUDA
-    // ABI element types such as cuDoubleComplex or cuComplex.
+    // Use ptr_as<T>() for host code and raw memory copies. For complex tensors, T is
+    // std::complex<...>. Use gpu_ptr_as<T>() for CUDA kernel code; for complex tensors, T is
+    // cuda::std::complex<...>. CUDA library APIs that require cuComplex ABI pointers should cast
+    // explicitly at the library-call boundary.
     template <typename T>
     T *ptr_as() const {
       cytnx_error_msg(this->dtype() != Type_class::cy_typeid_v<std::remove_cv_t<T>>,
@@ -535,10 +536,11 @@ namespace cytnx {
     // convert this->_impl->_storage->Mem to a typed variant of pointers, excluding void*
     gpu_pointer_types gpu_ptr() const;
 
-    // Convert storage to the CUDA ABI element pointer type.
+    // Convert storage to the CUDA kernel element pointer type.
     //
-    // This is for CUDA/cuBLAS/cuSOLVER/kernel interfaces. Generic host code and cudaMemcpy byte
-    // copies should use ptr_as<T>() so the logical Cytnx dtype mapping is checked.
+    // This is for CUDA kernel interfaces. Generic host code should use ptr_as<T>() so the logical
+    // Cytnx dtype mapping is checked. CUDA library APIs that require cuComplex ABI pointers should
+    // cast explicitly at the library-call boundary.
     template <typename T>
     T *gpu_ptr_as() const {
       cytnx_error_msg(

--- a/include/Type.hpp
+++ b/include/Type.hpp
@@ -13,6 +13,10 @@
 
 #include "cytnx_error.hpp"  // also brings in cuComplex.h
 
+#ifdef UNI_GPU
+  #include <cuda/std/complex>
+#endif
+
 #define MKL_Complex8 std::complex<float>
 #define MKL_Complex16 std::complex<double>
 
@@ -51,6 +55,11 @@ namespace cytnx {
   typedef std::complex<double> cytnx_complex128;
   typedef bool cytnx_bool;
 
+#ifdef UNI_GPU
+  using cytnx_cuda_complex64 = cuda::std::complex<float>;
+  using cytnx_cuda_complex128 = cuda::std::complex<double>;
+#endif
+
   namespace internal {
     template <class>
     struct is_complex_impl : std::false_type {};
@@ -58,11 +67,21 @@ namespace cytnx {
     template <class T>
     struct is_complex_impl<std::complex<T>> : std::true_type {};
 
+#ifdef UNI_GPU
+    template <class T>
+    struct is_complex_impl<cuda::std::complex<T>> : std::true_type {};
+#endif
+
     template <typename>
     struct is_complex_floating_point_impl : std::false_type {};
 
     template <typename T>
     struct is_complex_floating_point_impl<std::complex<T>> : std::is_floating_point<T> {};
+
+#ifdef UNI_GPU
+    template <typename T>
+    struct is_complex_floating_point_impl<cuda::std::complex<T>> : std::is_floating_point<T> {};
+#endif
 
     template <std::size_t I, typename T, typename Tuple>
     constexpr std::size_t index_in_tuple_helper() {
@@ -157,11 +176,13 @@ namespace cytnx {
     std::variant<void, cytnx_complex128, cytnx_complex64, cytnx_double, cytnx_float, cytnx_int64,
                  cytnx_uint64, cytnx_int32, cytnx_uint32, cytnx_int16, cytnx_uint16, cytnx_bool>;
 
-  // For GPU storage, the types are slightly different because CUDA uses their own complex type
+  // For GPU kernels, use cuda::std::complex for complex arithmetic. Low-level CUDA library calls
+  // that require cuComplex ABI pointers should cast explicitly at those call boundaries.
 #ifdef UNI_GPU
   using Type_list_gpu =
-    std::variant<void, cuDoubleComplex, cuComplex, cytnx_double, cytnx_float, cytnx_int64,
-                 cytnx_uint64, cytnx_int32, cytnx_uint32, cytnx_int16, cytnx_uint16, cytnx_bool>;
+    std::variant<void, cytnx_cuda_complex128, cytnx_cuda_complex64, cytnx_double, cytnx_float,
+                 cytnx_int64, cytnx_uint64, cytnx_int32, cytnx_uint32, cytnx_int16, cytnx_uint16,
+                 cytnx_bool>;
 #endif
 
   // The number of supported types

--- a/src/backend/algo_internal_gpu/cuSort_internal.cuh
+++ b/src/backend/algo_internal_gpu/cuSort_internal.cuh
@@ -20,7 +20,7 @@ namespace cytnx {
     template <typename T>
     __host__ __device__ bool cuSortCompare(const T& a, const T& b) {
       if constexpr (is_complex_v<T>) {
-        return (real(a) == real(b)) ? (imag(a) < imag(b)) : (real(a) < real(b));
+        return (a.real() == b.real()) ? (a.imag() < b.imag()) : (a.real() < b.real());
       } else if constexpr (Vec2Like<T>) {
         return (a.x == b.x) ? (a.y < b.y) : (a.x < b.x);
       } else {

--- a/src/backend/linalg_internal_gpu/cuAdd_internal.cu
+++ b/src/backend/linalg_internal_gpu/cuAdd_internal.cu
@@ -7,38 +7,9 @@ namespace cytnx {
 
     namespace {
 
-      template <typename T>
-      __device__ inline cuDoubleComplex CuToComplexDouble(const T &v) {
-        return make_cuDoubleComplex(static_cast<cytnx_double>(v), 0.0);
-      }
-
-      __device__ inline cuDoubleComplex CuToComplexDouble(const cuDoubleComplex &v) { return v; }
-
-      __device__ inline cuDoubleComplex CuToComplexDouble(const cuComplex &v) {
-        return cuComplexFloatToDouble(v);
-      }
-
-      template <typename T>
-      __device__ inline cuComplex CuToComplexFloat(const T &v) {
-        return make_cuFloatComplex(static_cast<cytnx_float>(v), 0.0f);
-      }
-
-      __device__ inline cuComplex CuToComplexFloat(const cuComplex &v) { return v; }
-
-      __device__ inline cuComplex CuToComplexFloat(const cuDoubleComplex &v) {
-        return make_cuFloatComplex(static_cast<cytnx_float>(cuCreal(v)),
-                                   static_cast<cytnx_float>(cuCimag(v)));
-      }
-
       template <typename TO, typename TL, typename TR>
       __device__ inline TO CuAddDispatchOp(const TL &lhs, const TR &rhs) {
-        if constexpr (std::is_same_v<TO, cuDoubleComplex>) {
-          return cuCadd(CuToComplexDouble(lhs), CuToComplexDouble(rhs));
-        } else if constexpr (std::is_same_v<TO, cuComplex>) {
-          return cuCaddf(CuToComplexFloat(lhs), CuToComplexFloat(rhs));
-        } else {
-          return static_cast<TO>(lhs) + static_cast<TO>(rhs);
-        }
+        return TO(lhs) + TO(rhs);
       }
 
       template <typename TO, typename TL, typename TR>
@@ -178,12 +149,12 @@ namespace cytnx {
                               const std::vector<cytnx_uint64> &invmapper_R) {
         switch (Rin->dtype()) {
           case Type.ComplexDouble:
-            cuAdd_dispatch_typed<TL, cuDoubleComplex>(out, Lin, Rin, len, shape, invmapper_L,
-                                                      invmapper_R);
+            cuAdd_dispatch_typed<TL, cytnx_cuda_complex128>(out, Lin, Rin, len, shape, invmapper_L,
+                                                            invmapper_R);
             break;
           case Type.ComplexFloat:
-            cuAdd_dispatch_typed<TL, cuComplex>(out, Lin, Rin, len, shape, invmapper_L,
-                                                invmapper_R);
+            cuAdd_dispatch_typed<TL, cytnx_cuda_complex64>(out, Lin, Rin, len, shape, invmapper_L,
+                                                           invmapper_R);
             break;
           case Type.Double:
             cuAdd_dispatch_typed<TL, cytnx_double>(out, Lin, Rin, len, shape, invmapper_L,
@@ -242,10 +213,12 @@ namespace cytnx {
 
       switch (Lin->dtype()) {
         case Type.ComplexDouble:
-          cuAdd_dispatch_rhs<cuDoubleComplex>(out, Lin, Rin, len, shape, invmapper_L, invmapper_R);
+          cuAdd_dispatch_rhs<cytnx_cuda_complex128>(out, Lin, Rin, len, shape, invmapper_L,
+                                                    invmapper_R);
           break;
         case Type.ComplexFloat:
-          cuAdd_dispatch_rhs<cuComplex>(out, Lin, Rin, len, shape, invmapper_L, invmapper_R);
+          cuAdd_dispatch_rhs<cytnx_cuda_complex64>(out, Lin, Rin, len, shape, invmapper_L,
+                                                   invmapper_R);
           break;
         case Type.Double:
           cuAdd_dispatch_rhs<cytnx_double>(out, Lin, Rin, len, shape, invmapper_L, invmapper_R);

--- a/src/backend/linalg_internal_gpu/cuDiv_dispatch.cu
+++ b/src/backend/linalg_internal_gpu/cuDiv_dispatch.cu
@@ -7,38 +7,9 @@ namespace cytnx {
 
     namespace {
 
-      template <typename T>
-      __device__ inline cuDoubleComplex CuToComplexDouble(const T &v) {
-        return make_cuDoubleComplex(static_cast<cytnx_double>(v), 0.0);
-      }
-
-      __device__ inline cuDoubleComplex CuToComplexDouble(const cuDoubleComplex &v) { return v; }
-
-      __device__ inline cuDoubleComplex CuToComplexDouble(const cuComplex &v) {
-        return cuComplexFloatToDouble(v);
-      }
-
-      template <typename T>
-      __device__ inline cuComplex CuToComplexFloat(const T &v) {
-        return make_cuFloatComplex(static_cast<cytnx_float>(v), 0.0f);
-      }
-
-      __device__ inline cuComplex CuToComplexFloat(const cuComplex &v) { return v; }
-
-      __device__ inline cuComplex CuToComplexFloat(const cuDoubleComplex &v) {
-        return make_cuFloatComplex(static_cast<cytnx_float>(cuCreal(v)),
-                                   static_cast<cytnx_float>(cuCimag(v)));
-      }
-
       template <typename TO, typename TL, typename TR>
       __device__ inline TO CuDivDispatchOp(const TL &lhs, const TR &rhs) {
-        if constexpr (std::is_same_v<TO, cuDoubleComplex>) {
-          return cuCdiv(CuToComplexDouble(lhs), CuToComplexDouble(rhs));
-        } else if constexpr (std::is_same_v<TO, cuComplex>) {
-          return cuCdivf(CuToComplexFloat(lhs), CuToComplexFloat(rhs));
-        } else {
-          return static_cast<TO>(lhs) / static_cast<TO>(rhs);
-        }
+        return TO(lhs) / TO(rhs);
       }
 
       template <typename TO, typename TL, typename TR>
@@ -178,12 +149,12 @@ namespace cytnx {
                               const std::vector<cytnx_uint64> &invmapper_R) {
         switch (Rin->dtype()) {
           case Type.ComplexDouble:
-            cuDiv_dispatch_typed<TL, cuDoubleComplex>(out, Lin, Rin, len, shape, invmapper_L,
-                                                      invmapper_R);
+            cuDiv_dispatch_typed<TL, cytnx_cuda_complex128>(out, Lin, Rin, len, shape, invmapper_L,
+                                                            invmapper_R);
             break;
           case Type.ComplexFloat:
-            cuDiv_dispatch_typed<TL, cuComplex>(out, Lin, Rin, len, shape, invmapper_L,
-                                                invmapper_R);
+            cuDiv_dispatch_typed<TL, cytnx_cuda_complex64>(out, Lin, Rin, len, shape, invmapper_L,
+                                                           invmapper_R);
             break;
           case Type.Double:
             cuDiv_dispatch_typed<TL, cytnx_double>(out, Lin, Rin, len, shape, invmapper_L,
@@ -242,10 +213,12 @@ namespace cytnx {
 
       switch (Lin->dtype()) {
         case Type.ComplexDouble:
-          cuDiv_dispatch_rhs<cuDoubleComplex>(out, Lin, Rin, len, shape, invmapper_L, invmapper_R);
+          cuDiv_dispatch_rhs<cytnx_cuda_complex128>(out, Lin, Rin, len, shape, invmapper_L,
+                                                    invmapper_R);
           break;
         case Type.ComplexFloat:
-          cuDiv_dispatch_rhs<cuComplex>(out, Lin, Rin, len, shape, invmapper_L, invmapper_R);
+          cuDiv_dispatch_rhs<cytnx_cuda_complex64>(out, Lin, Rin, len, shape, invmapper_L,
+                                                   invmapper_R);
           break;
         case Type.Double:
           cuDiv_dispatch_rhs<cytnx_double>(out, Lin, Rin, len, shape, invmapper_L, invmapper_R);

--- a/src/backend/linalg_internal_gpu/cuKron_internal.cuh
+++ b/src/backend/linalg_internal_gpu/cuKron_internal.cuh
@@ -46,7 +46,7 @@ namespace cytnx {
           x += cytnx_uint64(tmp2 / info[offset2 + j]) * info[len_nsa + j];
           y += cytnx_uint64(tmp2 % info[offset2 + j]) * info[offset1 + j];
         }
-        out[blockIdx.x * blockDim.x + threadIdx.x] = Lin[x] * Rin[y];
+        out[blockIdx.x * blockDim.x + threadIdx.x] = TO(Lin[x]) * TO(Rin[y]);
       }
     }
 

--- a/src/backend/linalg_internal_gpu/cuMul_dispatch.cu
+++ b/src/backend/linalg_internal_gpu/cuMul_dispatch.cu
@@ -7,38 +7,9 @@ namespace cytnx {
 
     namespace {
 
-      template <typename T>
-      __device__ inline cuDoubleComplex CuToComplexDouble(const T &v) {
-        return make_cuDoubleComplex(static_cast<cytnx_double>(v), 0.0);
-      }
-
-      __device__ inline cuDoubleComplex CuToComplexDouble(const cuDoubleComplex &v) { return v; }
-
-      __device__ inline cuDoubleComplex CuToComplexDouble(const cuComplex &v) {
-        return cuComplexFloatToDouble(v);
-      }
-
-      template <typename T>
-      __device__ inline cuComplex CuToComplexFloat(const T &v) {
-        return make_cuFloatComplex(static_cast<cytnx_float>(v), 0.0f);
-      }
-
-      __device__ inline cuComplex CuToComplexFloat(const cuComplex &v) { return v; }
-
-      __device__ inline cuComplex CuToComplexFloat(const cuDoubleComplex &v) {
-        return make_cuFloatComplex(static_cast<cytnx_float>(cuCreal(v)),
-                                   static_cast<cytnx_float>(cuCimag(v)));
-      }
-
       template <typename TO, typename TL, typename TR>
       __device__ inline TO CuMulDispatchOp(const TL &lhs, const TR &rhs) {
-        if constexpr (std::is_same_v<TO, cuDoubleComplex>) {
-          return cuCmul(CuToComplexDouble(lhs), CuToComplexDouble(rhs));
-        } else if constexpr (std::is_same_v<TO, cuComplex>) {
-          return cuCmulf(CuToComplexFloat(lhs), CuToComplexFloat(rhs));
-        } else {
-          return static_cast<TO>(lhs) * static_cast<TO>(rhs);
-        }
+        return TO(lhs) * TO(rhs);
       }
 
       template <typename TO, typename TL, typename TR>
@@ -178,12 +149,12 @@ namespace cytnx {
                               const std::vector<cytnx_uint64> &invmapper_R) {
         switch (Rin->dtype()) {
           case Type.ComplexDouble:
-            cuMul_dispatch_typed<TL, cuDoubleComplex>(out, Lin, Rin, len, shape, invmapper_L,
-                                                      invmapper_R);
+            cuMul_dispatch_typed<TL, cytnx_cuda_complex128>(out, Lin, Rin, len, shape, invmapper_L,
+                                                            invmapper_R);
             break;
           case Type.ComplexFloat:
-            cuMul_dispatch_typed<TL, cuComplex>(out, Lin, Rin, len, shape, invmapper_L,
-                                                invmapper_R);
+            cuMul_dispatch_typed<TL, cytnx_cuda_complex64>(out, Lin, Rin, len, shape, invmapper_L,
+                                                           invmapper_R);
             break;
           case Type.Double:
             cuMul_dispatch_typed<TL, cytnx_double>(out, Lin, Rin, len, shape, invmapper_L,
@@ -242,10 +213,12 @@ namespace cytnx {
 
       switch (Lin->dtype()) {
         case Type.ComplexDouble:
-          cuMul_dispatch_rhs<cuDoubleComplex>(out, Lin, Rin, len, shape, invmapper_L, invmapper_R);
+          cuMul_dispatch_rhs<cytnx_cuda_complex128>(out, Lin, Rin, len, shape, invmapper_L,
+                                                    invmapper_R);
           break;
         case Type.ComplexFloat:
-          cuMul_dispatch_rhs<cuComplex>(out, Lin, Rin, len, shape, invmapper_L, invmapper_R);
+          cuMul_dispatch_rhs<cytnx_cuda_complex64>(out, Lin, Rin, len, shape, invmapper_L,
+                                                   invmapper_R);
           break;
         case Type.Double:
           cuMul_dispatch_rhs<cytnx_double>(out, Lin, Rin, len, shape, invmapper_L, invmapper_R);

--- a/src/backend/linalg_internal_gpu/cuSub_dispatch.cu
+++ b/src/backend/linalg_internal_gpu/cuSub_dispatch.cu
@@ -7,38 +7,9 @@ namespace cytnx {
 
     namespace {
 
-      template <typename T>
-      __device__ inline cuDoubleComplex CuToComplexDouble(const T &v) {
-        return make_cuDoubleComplex(static_cast<cytnx_double>(v), 0.0);
-      }
-
-      __device__ inline cuDoubleComplex CuToComplexDouble(const cuDoubleComplex &v) { return v; }
-
-      __device__ inline cuDoubleComplex CuToComplexDouble(const cuComplex &v) {
-        return cuComplexFloatToDouble(v);
-      }
-
-      template <typename T>
-      __device__ inline cuComplex CuToComplexFloat(const T &v) {
-        return make_cuFloatComplex(static_cast<cytnx_float>(v), 0.0f);
-      }
-
-      __device__ inline cuComplex CuToComplexFloat(const cuComplex &v) { return v; }
-
-      __device__ inline cuComplex CuToComplexFloat(const cuDoubleComplex &v) {
-        return make_cuFloatComplex(static_cast<cytnx_float>(cuCreal(v)),
-                                   static_cast<cytnx_float>(cuCimag(v)));
-      }
-
       template <typename TO, typename TL, typename TR>
       __device__ inline TO CuSubDispatchOp(const TL &lhs, const TR &rhs) {
-        if constexpr (std::is_same_v<TO, cuDoubleComplex>) {
-          return cuCsub(CuToComplexDouble(lhs), CuToComplexDouble(rhs));
-        } else if constexpr (std::is_same_v<TO, cuComplex>) {
-          return cuCsubf(CuToComplexFloat(lhs), CuToComplexFloat(rhs));
-        } else {
-          return static_cast<TO>(lhs) - static_cast<TO>(rhs);
-        }
+        return TO(lhs) - TO(rhs);
       }
 
       template <typename TO, typename TL, typename TR>
@@ -178,12 +149,12 @@ namespace cytnx {
                               const std::vector<cytnx_uint64> &invmapper_R) {
         switch (Rin->dtype()) {
           case Type.ComplexDouble:
-            cuSub_dispatch_typed<TL, cuDoubleComplex>(out, Lin, Rin, len, shape, invmapper_L,
-                                                      invmapper_R);
+            cuSub_dispatch_typed<TL, cytnx_cuda_complex128>(out, Lin, Rin, len, shape, invmapper_L,
+                                                            invmapper_R);
             break;
           case Type.ComplexFloat:
-            cuSub_dispatch_typed<TL, cuComplex>(out, Lin, Rin, len, shape, invmapper_L,
-                                                invmapper_R);
+            cuSub_dispatch_typed<TL, cytnx_cuda_complex64>(out, Lin, Rin, len, shape, invmapper_L,
+                                                           invmapper_R);
             break;
           case Type.Double:
             cuSub_dispatch_typed<TL, cytnx_double>(out, Lin, Rin, len, shape, invmapper_L,
@@ -242,10 +213,12 @@ namespace cytnx {
 
       switch (Lin->dtype()) {
         case Type.ComplexDouble:
-          cuSub_dispatch_rhs<cuDoubleComplex>(out, Lin, Rin, len, shape, invmapper_L, invmapper_R);
+          cuSub_dispatch_rhs<cytnx_cuda_complex128>(out, Lin, Rin, len, shape, invmapper_L,
+                                                    invmapper_R);
           break;
         case Type.ComplexFloat:
-          cuSub_dispatch_rhs<cuComplex>(out, Lin, Rin, len, shape, invmapper_L, invmapper_R);
+          cuSub_dispatch_rhs<cytnx_cuda_complex64>(out, Lin, Rin, len, shape, invmapper_L,
+                                                   invmapper_R);
           break;
         case Type.Double:
           cuSub_dispatch_rhs<cytnx_double>(out, Lin, Rin, len, shape, invmapper_L, invmapper_R);

--- a/tests/Type_test.cpp
+++ b/tests/Type_test.cpp
@@ -14,6 +14,11 @@ namespace {
   using cytnx::cytnx_double;
   using cytnx::cytnx_int64;
   using cytnx::Type_list;
+#ifdef UNI_GPU
+  using cytnx::cytnx_cuda_complex128;
+  using cytnx::cytnx_cuda_complex64;
+  using cytnx::Type_list_gpu;
+#endif
   using cytnx::variant_contains_v;
   using cytnx::variant_index_v;
 
@@ -37,6 +42,21 @@ namespace {
   static_assert(cytnx::Type_class::Uint64 == cytnx::Type_class::Int64 + 1);
   static_assert(cytnx::Type_class::Uint32 == cytnx::Type_class::Int32 + 1);
   static_assert(cytnx::Type_class::Uint16 == cytnx::Type_class::Int16 + 1);
+
+#ifdef UNI_GPU
+  static_assert(variant_index_v<cytnx_cuda_complex128, Type_list_gpu> ==
+                cytnx::Type_class::ComplexDouble);
+  static_assert(variant_index_v<cytnx_cuda_complex64, Type_list_gpu> ==
+                cytnx::Type_class::ComplexFloat);
+  static_assert(cytnx::is_complex_v<cytnx_cuda_complex128>);
+  static_assert(cytnx::is_complex_v<cytnx_cuda_complex64>);
+  static_assert(
+    std::is_same_v<cytnx::Type_class::type_promote_gpu_t<cytnx_cuda_complex64, cytnx_double>,
+                   cytnx_cuda_complex128>);
+  static_assert(
+    std::is_same_v<cytnx::Type_class::type_promote_gpu_t<cytnx_double, cytnx_cuda_complex64>,
+                   cytnx_cuda_complex128>);
+#endif
 
 }  // namespace
 

